File: leetcode-stuff/dfs/count_province.cpp
#include <vector>
#include <cstring>
#include <functional>

class Solution {

    public:
        int findCircleNum(std::vector<std::vector<int>>& isConnected) // note: & is used for referencing for saving and update
        {
            // Get the number of cities(nodes)
            int num_cities = isConnected.size();

            // Initialize the count of provinces
            int province_count = 0;

            // Initialize the visited array to keep track of the visited cities for the sequence of iterations
            bool visited[num_cities]; // reminder: been too long, this is an array
            // the reason we are also including vector library since vector has more functions to use

            // Initialize all cities as unvisited.
            std::memset(visited, false, sizeof(visited));

            // Define the depth-first search algorithm (DFS) as a lambda function
            std::function<void(int)> dfs = [&](int cityIndex){
                // Mark the current city as visited 
                visited[cityIndex] = true;

                // Visit all the unvisited cities if there is a connnection to that city with city of cityIndex
                for (int j=0; j< num_cities; j++){
                    if (!visited[j] && isConnected[cityIndex][j]){
                        dfs(j);
                    }
                }
            };
                // Iterate over each city to count the number of provinces.
            for (int i = 0; i < num_cities; ++i) {
                // If the city is not yet visited, it is part of a new province.
                if (!visited[i]) {
                    dfs(i); // Perform DFS to visit all cities in the current province.
                    ++province_count; // Increment the count of provinces.
                }
            }
        
            // Return the total number of provinces found.
            return province_count;
        }
};

File: leetcode-stuff/dfs/count_province.py
# Source for reference: https://algo.monster/liteproblems/547

from typing import List

class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        # Depth-First Search function which marks the nodes as visited
        def dfs(current_city: int):
            visited[current_city] = True  # Mark the current city as visited
            for adjacent_city, connected in enumerate(isConnected[current_city]):
                # If the adjacent city is not visited and there is a connection,
                # then continue the search from that city
                if not visited[adjacent_city] and connected:
                    dfs(adjacent_city)

        # Number of cities in the given matrix
        num_cities = len(isConnected)
        # Initialize a visited list to keep track of cities that have been visited
        visited = [False] * num_cities
        # Counter for the number of provinces (disconnected components)
        province_count = 0
        # Loop over each city and perform DFS if it hasn't been visited
        for city in range(num_cities):
            # If the city hasn't been visited yet
            # basically any unvisited cities after we have defined all the relevant connected cities
            # will be seen as a new set of province, starting with that unvisited city
            if not visited[city]:  
                dfs(city)  # Start DFS from this city
                # After finishing DFS, we have found a new province
                province_count += 1
        # Return the total number of disconnected components (provinces) in the graph
        return province_count
        